src/settings.rs

```diff
@@ -26,7 +30,226 @@ impl Settings {
 }
 
 impl RepositorySettings {
-    pub fn closed_since_date(&self) -> DateTime<Utc> {
-        Utc::now()
+    pub fn closed_since_date(&self) -> Option<DateTime<Utc>> {
+        if let Some(ref s) = self.since {
+            match *s {
+                Since { quantity: q @ _, unit: SinceSpan::Hour } => {
+                    return Some(Utc::now().checked_sub_signed(Duration::hours(q as i64)).unwrap())
+                },
+                Since { quantity: q @ _, unit: SinceSpan::Day } => {
+                    return Some(Utc::now().checked_sub_signed(Duration::days(q as i64)).unwrap())
+                },
+                Since { quantity: q @ _, unit: SinceSpan::Week } => {
+                    return Some(Utc::now().checked_sub_signed(Duration::weeks(q as i64)).unwrap())
+                },
+            }
+        };
+
+        None
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Since {
+    quantity: u64,
+    unit: SinceSpan,
+}
+
+impl FromStr for Since {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let tokens: Vec<&str> = s.split_whitespace().collect();
+        if tokens.len() != 2 {
+            return Err(format_err!("'{}' should be exactly two words (found {})", s, tokens.len()));
+        }
+
+        let quantity = tokens[0].parse::<u64>()?;
+        let interval = tokens[1].parse::<SinceSpan>()?;
+
+        Ok(Since { quantity: quantity, unit: interval, })
+    }
+}
+
+impl<'de> Deserialize<'de> for Since {
```

Comment:
A few questions here

what does `'de` do in this case?

where does `deserialize` get used? I see it in a comment above on line 15, but I'm not sure how it's actually being used

Reply:
`'de` is a [lifetime annotation](https://doc.rust-lang.org/book/first-edition/lifetimes.html#impl-blocks). The usage of it on this line is basically defining the lifetime so it can be used on line 76. There isn't really anything special going on with lifetimes here, IIRC we have to specify the lifetime because of how the `Deserialize` trait itself is defined.

The reference to the `Deserialize` trait on line 15 isn't actually a comment. It's an attribute asking the compiler to automatically generate an `impl <'de> Deserialize<'de> for RepositorySettings` block, and it's able to do that because everything in `RepositorySettings` already implements the `Deserialize` trait.

For where `deserialize` actually gets used: The `config` crate uses `serde` (which is where the `Deserialize` trait comes from) to do the `YAML -> struct` conversion. We're using the `config` crate to do this in [the new method of the Settings struct](https://github.com/jhelwig/gh-univiewer/pull/9/files#diff-7e30a274b7e58a9d8bf644a467c7e044R28).